<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 4 Session 4 | Introduction to Base R</title>
  <meta name="description" content="Introductory tutorial to base R, with a biological focus. In particular it will focus on analysing RNA-sequencing data with Bioconductor." />
  <meta name="generator" content="bookdown 0.19 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 4 Session 4 | Introduction to Base R" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Introductory tutorial to base R, with a biological focus. In particular it will focus on analysing RNA-sequencing data with Bioconductor." />
  <meta name="github-repo" content="shians/BaseR_Intro" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 4 Session 4 | Introduction to Base R" />
  
  <meta name="twitter:description" content="Introductory tutorial to base R, with a biological focus. In particular it will focus on analysing RNA-sequencing data with Bioconductor." />
  

<meta name="author" content="" />


<meta name="date" content="2020-10-30" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="session-3.html"/>
<link rel="next" href="further-reading.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/kePrint-0.0.1/kePrint.js"></script>


<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Introduction to Base R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a><ul>
<li class="chapter" data-level="0.1" data-path="index.html"><a href="index.html#data-files"><i class="fa fa-check"></i><b>0.1</b> Data files</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="session-1.html"><a href="session-1.html"><i class="fa fa-check"></i><b>1</b> Session 1</a><ul>
<li class="chapter" data-level="1.1" data-path="session-1.html"><a href="session-1.html#r-and-rstudio"><i class="fa fa-check"></i><b>1.1</b> R and RStudio</a></li>
<li class="chapter" data-level="1.2" data-path="session-1.html"><a href="session-1.html#r-scripts"><i class="fa fa-check"></i><b>1.2</b> R scripts</a></li>
<li class="chapter" data-level="1.3" data-path="session-1.html"><a href="session-1.html#working-directory"><i class="fa fa-check"></i><b>1.3</b> Working directory</a></li>
<li class="chapter" data-level="1.4" data-path="session-1.html"><a href="session-1.html#maths"><i class="fa fa-check"></i><b>1.4</b> Maths</a></li>
<li class="chapter" data-level="1.5" data-path="session-1.html"><a href="session-1.html#comps"><i class="fa fa-check"></i><b>1.5</b> Comparisons</a></li>
<li class="chapter" data-level="1.6" data-path="session-1.html"><a href="session-1.html#functions"><i class="fa fa-check"></i><b>1.6</b> Functions</a></li>
<li class="chapter" data-level="1.7" data-path="session-1.html"><a href="session-1.html#variables"><i class="fa fa-check"></i><b>1.7</b> Variables</a></li>
<li class="chapter" data-level="1.8" data-path="session-1.html"><a href="session-1.html#data-types"><i class="fa fa-check"></i><b>1.8</b> Data types</a><ul>
<li class="chapter" data-level="1.8.1" data-path="session-1.html"><a href="session-1.html#factors"><i class="fa fa-check"></i><b>1.8.1</b> Factors</a></li>
</ul></li>
<li class="chapter" data-level="1.9" data-path="session-1.html"><a href="session-1.html#data-structures"><i class="fa fa-check"></i><b>1.9</b> Data structures</a><ul>
<li class="chapter" data-level="1.9.1" data-path="session-1.html"><a href="session-1.html#vector"><i class="fa fa-check"></i><b>1.9.1</b> Vector</a></li>
<li class="chapter" data-level="1.9.2" data-path="session-1.html"><a href="session-1.html#matrix"><i class="fa fa-check"></i><b>1.9.2</b> Matrix</a></li>
<li class="chapter" data-level="1.9.3" data-path="session-1.html"><a href="session-1.html#data-frame"><i class="fa fa-check"></i><b>1.9.3</b> Data Frame</a></li>
</ul></li>
<li class="chapter" data-level="1.10" data-path="session-1.html"><a href="session-1.html#reading-in-data"><i class="fa fa-check"></i><b>1.10</b> Reading in data</a></li>
<li class="chapter" data-level="1.11" data-path="session-1.html"><a href="session-1.html#vect"><i class="fa fa-check"></i><b>1.11</b> Vectorised Operations</a></li>
<li class="chapter" data-level="1.12" data-path="session-1.html"><a href="session-1.html#write"><i class="fa fa-check"></i><b>1.12</b> Writing out data</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="session-2.html"><a href="session-2.html"><i class="fa fa-check"></i><b>2</b> Session 2</a><ul>
<li class="chapter" data-level="2.1" data-path="session-2.html"><a href="session-2.html#miss"><i class="fa fa-check"></i><b>2.1</b> Missing values</a></li>
<li class="chapter" data-level="2.2" data-path="session-2.html"><a href="session-2.html#subsetting-vectors"><i class="fa fa-check"></i><b>2.2</b> Subsetting vectors</a><ul>
<li class="chapter" data-level="2.2.1" data-path="session-2.html"><a href="session-2.html#subsetting-with-numeric-indicies"><i class="fa fa-check"></i><b>2.2.1</b> Subsetting with numeric indicies</a></li>
<li class="chapter" data-level="2.2.2" data-path="session-2.html"><a href="session-2.html#subsetting-with-logical-vectors"><i class="fa fa-check"></i><b>2.2.2</b> Subsetting with logical vectors</a></li>
<li class="chapter" data-level="2.2.3" data-path="session-2.html"><a href="session-2.html#useful-summaries-of-logical-vectors"><i class="fa fa-check"></i><b>2.2.3</b> Useful summaries of logical vectors</a></li>
<li class="chapter" data-level="2.2.4" data-path="session-2.html"><a href="session-2.html#overwriting-subset-values"><i class="fa fa-check"></i><b>2.2.4</b> Overwriting subset values</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="session-2.html"><a href="session-2.html#subsetting-data-structures"><i class="fa fa-check"></i><b>2.3</b> Subsetting data structures</a><ul>
<li class="chapter" data-level="2.3.1" data-path="session-2.html"><a href="session-2.html#subsetting-data-structures-with-numerics"><i class="fa fa-check"></i><b>2.3.1</b> Subsetting data structures with numerics</a></li>
<li class="chapter" data-level="2.3.2" data-path="session-2.html"><a href="session-2.html#subsettting-data-frames-using-characters"><i class="fa fa-check"></i><b>2.3.2</b> Subsettting data frames using characters</a></li>
<li class="chapter" data-level="2.3.3" data-path="session-2.html"><a href="session-2.html#removing-na-rows-using-complete.cases"><i class="fa fa-check"></i><b>2.3.3</b> Removing NA rows using complete.cases()</a></li>
<li class="chapter" data-level="2.3.4" data-path="session-2.html"><a href="session-2.html#rearranging-using-the-order-function"><i class="fa fa-check"></i><b>2.3.4</b> Rearranging using the <code>order()</code> function</a></li>
<li class="chapter" data-level="2.3.5" data-path="session-2.html"><a href="session-2.html#subsetting-data-structures-with-logicals"><i class="fa fa-check"></i><b>2.3.5</b> Subsetting data structures with logicals</a></li>
<li class="chapter" data-level="2.3.6" data-path="session-2.html"><a href="session-2.html#in-and"><i class="fa fa-check"></i><b>2.3.6</b> %in%, &amp; and |</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="session-2.html"><a href="session-2.html#merge"><i class="fa fa-check"></i><b>2.4</b> Merge</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="session-3.html"><a href="session-3.html"><i class="fa fa-check"></i><b>3</b> Session 3</a><ul>
<li class="chapter" data-level="3.1" data-path="session-3.html"><a href="session-3.html#basic-plotting"><i class="fa fa-check"></i><b>3.1</b> Basic plotting</a><ul>
<li class="chapter" data-level="3.1.1" data-path="session-3.html"><a href="session-3.html#scatter-plots"><i class="fa fa-check"></i><b>3.1.1</b> Scatter plots</a></li>
<li class="chapter" data-level="3.1.2" data-path="session-3.html"><a href="session-3.html#histograms"><i class="fa fa-check"></i><b>3.1.2</b> Histograms</a></li>
<li class="chapter" data-level="3.1.3" data-path="session-3.html"><a href="session-3.html#boxplots"><i class="fa fa-check"></i><b>3.1.3</b> Boxplots</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="session-3.html"><a href="session-3.html#lists"><i class="fa fa-check"></i><b>3.2</b> Lists</a><ul>
<li class="chapter" data-level="3.2.1" data-path="session-3.html"><a href="session-3.html#indexing-lists"><i class="fa fa-check"></i><b>3.2.1</b> Indexing lists</a></li>
<li class="chapter" data-level="3.2.2" data-path="session-3.html"><a href="session-3.html#chaining-list-indices"><i class="fa fa-check"></i><b>3.2.2</b> Chaining list indices</a></li>
<li class="chapter" data-level="3.2.3" data-path="session-3.html"><a href="session-3.html#named-list"><i class="fa fa-check"></i><b>3.2.3</b> Named list</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="session-3.html"><a href="session-3.html#classes"><i class="fa fa-check"></i><b>3.3</b> Classes</a></li>
<li class="chapter" data-level="3.4" data-path="session-3.html"><a href="session-3.html#packages"><i class="fa fa-check"></i><b>3.4</b> Packages</a></li>
<li class="chapter" data-level="3.5" data-path="session-3.html"><a href="session-3.html#dgelist"><i class="fa fa-check"></i><b>3.5</b> DGEList</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="session-4.html"><a href="session-4.html"><i class="fa fa-check"></i><b>4</b> Session 4</a><ul>
<li class="chapter" data-level="4.1" data-path="session-4.html"><a href="session-4.html#dgelist-1"><i class="fa fa-check"></i><b>4.1</b> DGEList</a></li>
<li class="chapter" data-level="4.2" data-path="session-4.html"><a href="session-4.html#filtering"><i class="fa fa-check"></i><b>4.2</b> Filtering</a></li>
<li class="chapter" data-level="4.3" data-path="session-4.html"><a href="session-4.html#normalisation"><i class="fa fa-check"></i><b>4.3</b> Normalisation</a></li>
<li class="chapter" data-level="4.4" data-path="session-4.html"><a href="session-4.html#mds-plots"><i class="fa fa-check"></i><b>4.4</b> MDS plots</a></li>
<li class="chapter" data-level="4.5" data-path="session-4.html"><a href="session-4.html#linear-modelling"><i class="fa fa-check"></i><b>4.5</b> Linear modelling</a><ul>
<li class="chapter" data-level="4.5.1" data-path="session-4.html"><a href="session-4.html#design-matrix"><i class="fa fa-check"></i><b>4.5.1</b> Design matrix</a></li>
<li class="chapter" data-level="4.5.2" data-path="session-4.html"><a href="session-4.html#contrasts"><i class="fa fa-check"></i><b>4.5.2</b> Contrasts</a></li>
<li class="chapter" data-level="4.5.3" data-path="session-4.html"><a href="session-4.html#variance-modelling-with-voom"><i class="fa fa-check"></i><b>4.5.3</b> Variance modelling with voom</a></li>
<li class="chapter" data-level="4.5.4" data-path="session-4.html"><a href="session-4.html#fitting-the-linear-model"><i class="fa fa-check"></i><b>4.5.4</b> Fitting the linear model</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="session-4.html"><a href="session-4.html#statistical-testing"><i class="fa fa-check"></i><b>4.6</b> Statistical testing</a></li>
<li class="chapter" data-level="4.7" data-path="session-4.html"><a href="session-4.html#ma-plot"><i class="fa fa-check"></i><b>4.7</b> MA Plot</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="further-reading.html"><a href="further-reading.html"><i class="fa fa-check"></i><b>5</b> Further Reading</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Introduction to Base R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="session-4" class="section level1">
<h1><span class="header-section-number">Chapter 4</span> Session 4</h1>
<p>In this session we will run through the basic steps for analysing a simple
RNA-seq experiment using the
<a href="https://f1000research.com/articles/5-1408">limma-voom workflow</a>. This includes:</p>
<ul>
<li>filtering out lowly expressed genes</li>
<li>normalisation</li>
<li>creating a multidimensional scaling (MDS) plot</li>
<li>creating a design matrix</li>
<li>fitting gene-wise linear models (with empirical Bayes moderation to more
accurately estimate gene-wise variability)</li>
<li>performing statistical testing for differential expression</li>
</ul>
<p>The aim of this session is to give you experience with a real-world RNA-seq
analysis, and making extensive use of an external library. We will not cover the
statistics in any depth. In general analysis packages will want your data in
some specific format, so it’s important to be able to manipulate the data to fit
the package’s requirements.</p>
<p>Much of the materials here are explained in greater detail in the limma user’s
guide. You can view this by typing <code>help("limma")</code> and following the links.</p>
<div id="dgelist-1" class="section level2">
<h2><span class="header-section-number">4.1</span> DGEList</h2>
<p>The data we are looking at comes from three cell populations (basal, luminal
progenitor (LP) and mature luminal (ML)) sorted from the mammary glands of
female virgin mice, each profiled in triplicate.</p>
<p>Let’s start by creating our <code>DGEList</code> object. As a reminder, this object
contains 3 key pieces of data:</p>
<ul>
<li><code>counts</code>: the main data of this object, a matrix of count values with samples
along the columns and features/genes along the rows.</li>
<li><code>samples</code>: a data frame containing annotation for the samples. The rows
in this table describe the corresponding column of the counts data.</li>
<li><code>genes</code>: a data frame containing annotation for the genes in the counts
matrix. The rows in this table describe the corresponding row in the counts
matrix.</li>
</ul>
<div class="sourceCode" id="cb238"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb238-1"><a href="session-4.html#cb238-1"></a><span class="co"># load required packages</span></span>
<span id="cb238-2"><a href="session-4.html#cb238-2"></a><span class="kw">library</span>(edgeR)</span>
<span id="cb238-3"><a href="session-4.html#cb238-3"></a><span class="kw">library</span>(limma)</span>
<span id="cb238-4"><a href="session-4.html#cb238-4"></a></span>
<span id="cb238-5"><a href="session-4.html#cb238-5"></a><span class="co"># vector of file names</span></span>
<span id="cb238-6"><a href="session-4.html#cb238-6"></a>files &lt;-<span class="st"> </span><span class="kw">dir</span>(<span class="dt">path =</span> <span class="st">&quot;data/counts&quot;</span>, <span class="dt">pattern =</span> <span class="st">&quot;GSM&quot;</span>)</span>
<span id="cb238-7"><a href="session-4.html#cb238-7"></a>group &lt;-<span class="st"> </span><span class="kw">factor</span>(<span class="kw">c</span>(<span class="st">&quot;LP&quot;</span>, <span class="st">&quot;ML&quot;</span>, <span class="st">&quot;Basal&quot;</span>, <span class="st">&quot;Basal&quot;</span>, <span class="st">&quot;ML&quot;</span>, <span class="st">&quot;LP&quot;</span>, <span class="st">&quot;Basal&quot;</span>, <span class="st">&quot;ML&quot;</span>, <span class="st">&quot;LP&quot;</span>))</span>
<span id="cb238-8"><a href="session-4.html#cb238-8"></a>samplenames &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;10_6_5_11&quot;</span>, <span class="st">&quot;9_6_5_11&quot;</span>, <span class="st">&quot;purep53&quot;</span>, <span class="st">&quot;JMS8-2&quot;</span>, <span class="st">&quot;JMS8-3&quot;</span>, <span class="st">&quot;JMS8-4&quot;</span>, </span>
<span id="cb238-9"><a href="session-4.html#cb238-9"></a>    <span class="st">&quot;JMS8-5&quot;</span>, <span class="st">&quot;JMS9-P7c&quot;</span>, <span class="st">&quot;JMS9-P8c&quot;</span>)</span>
<span id="cb238-10"><a href="session-4.html#cb238-10"></a></span>
<span id="cb238-11"><a href="session-4.html#cb238-11"></a></span>
<span id="cb238-12"><a href="session-4.html#cb238-12"></a><span class="co"># create DGEList object</span></span>
<span id="cb238-13"><a href="session-4.html#cb238-13"></a>dge &lt;-<span class="st"> </span><span class="kw">readDGE</span>(files, <span class="dt">path =</span> <span class="st">&quot;data/counts&quot;</span>, <span class="dt">columns =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span>), <span class="dt">group =</span> group, <span class="dt">labels =</span> samplenames)</span>
<span id="cb238-14"><a href="session-4.html#cb238-14"></a></span>
<span id="cb238-15"><a href="session-4.html#cb238-15"></a><span class="co"># add gene annotation information</span></span>
<span id="cb238-16"><a href="session-4.html#cb238-16"></a>dge<span class="op">$</span>genes &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">&quot;data/Ses3_geneAnnot.tsv&quot;</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<p>You can index the DGEList object by treating it as if it were the counts matrix,
the object will handle the extraction and ordering of the sample and gene
annotation data frames.</p>
</div>
<div id="filtering" class="section level2">
<h2><span class="header-section-number">4.2</span> Filtering</h2>
<p>The first step is to filter out lowly expressed genes. There are two main
problems with low abundant genes:</p>
<ul>
<li>Technical variation is more problematic for low abundance genes. This
variation is thought to be due to two factors; insufficient mixing and
low sampling fraction <span class="citation">(McIntyre et al. <a href="#ref-mcintyre2011rna" role="doc-biblioref">2011</a>)</span>.
<ul>
<li>Insufficient mixing of solutions during library preparation can result
in uneven distribution of reads.</li>
<li>RNA sequencing can be thought of as sampling. Measurement errors will
occur simply due to the random nature of the sampling process. This problem
affects lowly abundant RNA species more because the relative error for small
count values is larger than it would be for more highly abundant RNA
species.</li>
</ul></li>
<li>Biologically, genes that are expressed at low, biologically not meaningful,
levels are not of interest.</li>
</ul>
<blockquote>
<p>Removing these highly variable, lowly expressed genes increases your ‘power’
to detect differentially expressed genes <span class="citation">(Bourgon, Gentleman, and Huber <a href="#ref-bourgon2010independent" role="doc-biblioref">2010</a>)</span>, where
‘power’ is your ability to detect true positives. In testing for differential
expression, a statistical test is conducted for each gene. When a high number
of statistical tests are performed, a portion of them will be significant
purely due to random chance. A common procedure to control for the number of
false positive is to perform ‘multiple testing correction’ on the p-values. This
adjusts the p-value in a way that reduces the number of false positives but
comes at the cost of reduced power to detect true positives. If we filter out
uninteresting, lowly expressed genes, we need to perform fewer statistical tests
and reduce the impact that multiple testing adjustment has on detection power.</p>
</blockquote>
<p>The <code>filterByExpr()</code> function provides an automatic way to filter genes.</p>
<p>Roughly speaking, by default, it keeps genes with a count of 10 or more, in at
least as many samples as the smallest experimental group. In our experiment,
there are 3 phenotype groups each with 3 samples. Therefore we retain only genes
that have 10 or more counts in 3 or more samples.</p>
<p>More specifically, the actual filtering is done on counts per million, with
similar result to the above criteria. This is to prevent bias against samples
with small library sizes.</p>
<p>The output of this function is a vector of logicals, indicating which genes
(rows) should be kept and which filtered.</p>
<div class="sourceCode" id="cb239"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb239-1"><a href="session-4.html#cb239-1"></a>keep &lt;-<span class="st"> </span><span class="kw">filterByExpr</span>(dge)</span>
<span id="cb239-2"><a href="session-4.html#cb239-2"></a><span class="kw">table</span>(keep)</span></code></pre></div>
<pre><code>## keep
## FALSE  TRUE 
## 10555 16624</code></pre>
<div class="sourceCode" id="cb241"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb241-1"><a href="session-4.html#cb241-1"></a>dge &lt;-<span class="st"> </span>dge[keep, , keep.lib.sizes =<span class="st"> </span><span class="ot">FALSE</span>]</span>
<span id="cb241-2"><a href="session-4.html#cb241-2"></a><span class="kw">dim</span>(dge<span class="op">$</span>counts)</span></code></pre></div>
<pre><code>## [1] 16624     9</code></pre>
<p>We can see that we now have 16624 genes. We started with 27179 genes - meaning
that ~40% of genes have been filtered out.</p>
</div>
<div id="normalisation" class="section level2">
<h2><span class="header-section-number">4.3</span> Normalisation</h2>
<p>The aim of normalisation is to remove systematic technical effects. There are
two main factors that need to be normalised for in RNA-seq:</p>
<ul>
<li>Sequencing depth/library size - technically, sequencing a sample to half the
depth will give, on average, half the number of reads mapping to each gene
<span class="citation">(Robinson and Oshlack <a href="#ref-robinson2010scaling" role="doc-biblioref">2010</a>)</span>.</li>
<li>RNA composition - if a large number of genes are unique to, or highly
expressed in, only one experimental condition, the sequencing capacity available
for the remaining genes in that sample is decreased. For example, if there are
only five genes being studied in two experimental groups, if one gene is
particularly high in group A, then with limited sequencing depth, that gene
will reduce the counts of the remaining four genes. The effect of this is that
the remaining four genes appear under-expressed in group A compared to
group B when the true amount of gene product is actually equal for these 4
genes <span class="citation">(Robinson and Oshlack <a href="#ref-robinson2010scaling" role="doc-biblioref">2010</a>)</span>.</li>
</ul>
<p>Sequencing depth is accounted for by calculating the counts per million (cpm).
This metric is calculated by:</p>
<ol style="list-style-type: decimal">
<li>taking the library size (sum of all counts for a sample),</li>
<li>dividing this by 1,000,000 to get the ‘per million’ scaling factor,</li>
<li>then dividing all read counts for each gene in that sample by the
‘per million’ scaling factor</li>
</ol>
<p>RNA composition can be accounted for by using more sophisticated normalisation
methodologies. We will use ‘trimmed mean of M-values’ (TMM), which estimates
relative RNA levels from RNA-seq data <span class="citation">(Robinson and Oshlack <a href="#ref-robinson2010scaling" role="doc-biblioref">2010</a>)</span>. Under the
assumption that most genes are not differentially expressed, TMM calculates a
library size scaling factor for each library (sample). This is done using the
following steps:</p>
<ol style="list-style-type: decimal">
<li>calculate the gene expression log fold changes and absolute expression values
for pair-wise samples (selecting one sample from the experiment as a reference)</li>
<li>remove the genes with the highest and lowest fold changes and absolute
expression values</li>
<li>take a weighted mean of the remaining genes (where the weight is the inverse
of the approximate asymptotic variances). This gives the normalisation factor
for each library (sample)</li>
</ol>
<p>Subsequent steps in this analysis will use log-cpm values, calculated using the
normalisation factors, which scales each library size.</p>
<p>We can calculate the normalisation factors, specifying that we want to use the
<code>"TMM"</code> method:</p>
<div class="sourceCode" id="cb243"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb243-1"><a href="session-4.html#cb243-1"></a>dge &lt;-<span class="st"> </span><span class="kw">calcNormFactors</span>(dge, <span class="dt">method =</span> <span class="st">&quot;TMM&quot;</span>)</span></code></pre></div>
<p>This function calculates the normalisation factors for each library (sample)
and puts this information in the <code>samples</code> data frame. Note that it takes dge
(our <code>DGEList</code> object as input) and returns a <code>DGEList</code> object as well.</p>
<p>Let’s take a look at our normalisation factors:</p>
<div class="sourceCode" id="cb244"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb244-1"><a href="session-4.html#cb244-1"></a>dge<span class="op">$</span>samples</span></code></pre></div>
<pre><code>##                              files group lib.size norm.factors
## 10_6_5_11 GSM1545535_10_6_5_11.txt    LP 32857304    0.8943956
## 9_6_5_11   GSM1545536_9_6_5_11.txt    ML 35328624    1.0250186
## purep53     GSM1545538_purep53.txt Basal 57147943    1.0459005
## JMS8-2       GSM1545539_JMS8-2.txt Basal 51356800    1.0458455
## JMS8-3       GSM1545540_JMS8-3.txt    ML 75782871    1.0162707
## JMS8-4       GSM1545541_JMS8-4.txt    LP 60506774    0.9217132
## JMS8-5       GSM1545542_JMS8-5.txt Basal 55073018    0.9961959
## JMS9-P7c   GSM1545544_JMS9-P7c.txt    ML 21305254    1.0861026
## JMS9-P8c   GSM1545545_JMS9-P8c.txt    LP 19955335    0.9839203</code></pre>
<p>These normalisation factors are all close to 1 for all samples, suggesting
minimal difference in relative RNA levels between samples.</p>
<div class="sourceCode" id="cb246"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb246-1"><a href="session-4.html#cb246-1"></a><span class="kw">boxplot</span>(<span class="kw">log</span>(dge<span class="op">$</span>counts <span class="op">+</span><span class="st"> </span><span class="fl">0.5</span>))</span></code></pre></div>
<p><img src="IntroBaseR_files/figure-html/unnamed-chunk-112-1.png" width="672" /></p>
<div class="sourceCode" id="cb247"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb247-1"><a href="session-4.html#cb247-1"></a><span class="kw">boxplot</span>(<span class="kw">cpm</span>(dge<span class="op">$</span>counts, <span class="dt">log =</span> <span class="ot">TRUE</span>))</span></code></pre></div>
<p><img src="IntroBaseR_files/figure-html/unnamed-chunk-113-1.png" width="672" /></p>
<p><strong>Challenge 4.1</strong></p>
<ol style="list-style-type: decimal">
<li><p>Create a boxplot of the normalisation factors versus group.</p></li>
<li><p><code>rowSums()</code> can be used on a matrix to take sums across the rows. Use this
to retrieve annotations of all genes with more than 1,000,000 total counts
across all samples.</p></li>
<li><p>What are the 5 most highly expressed genes within the ML samples?</p></li>
</ol>
</div>
<div id="mds-plots" class="section level2">
<h2><span class="header-section-number">4.4</span> MDS plots</h2>
<p>Before we perform statistical tests, it’s useful to perform some exploratory
visual analysis to get an overall idea of how our data is behaving.</p>
<p>MDS is a way to visualise distances between sets of data points
(samples in our case). It is a dimensionality reduction technique, similar to
principal components analysis (PCA). We treat gene expression in samples as if
they were coordinates in a high-dimensional coordinate system, then we can find
“distances” between samples as we do between points in space. Then the goal of
the algorithm is to find a representation in lower dimensional space such that
points that the distance of two objects from each other in high dimensional
space is preserved in lower dimensions.</p>
<p>The <code>plotMDS()</code> from <code>limma</code> creates an MDS plot from a <code>DGEList</code> object.</p>
<div class="sourceCode" id="cb248"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb248-1"><a href="session-4.html#cb248-1"></a><span class="kw">plotMDS</span>(dge)</span></code></pre></div>
<p><img src="IntroBaseR_files/figure-html/unnamed-chunk-114-1.png" width="672" /></p>
<p>Each point on the plot represents one sample and is ‘labelled’ using the sample
name. The distances between each sample in the resulting plot can be interpreted
as the typical log2-fold-change between the samples, for the most differentially
expressed genes.</p>
<p>We can change the labelling to use the name of the group the sample belongs to
instead:</p>
<div class="sourceCode" id="cb249"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb249-1"><a href="session-4.html#cb249-1"></a><span class="kw">plotMDS</span>(dge, <span class="dt">labels =</span> group)</span></code></pre></div>
<p><img src="IntroBaseR_files/figure-html/unnamed-chunk-115-1.png" width="672" /></p>
<p>This shows us that the phenotype groups tend to cluster together, meaning that
the gene expression profiles are similar for samples within a phenotype group.
The ‘Basal’ type samples quite close together while the ‘LP’
(luminal progenitor) and ‘ML’ (mature luminal) type samples are further apart,
signifying that their expression profiles are more variable.</p>
<p>To make the three phenotype groups more distinct in our graph, we are going to
colour samples from each group differently. To do this, we will use the <code>col</code>
argument in <code>plotMDS()</code>. <code>col</code> takes in a vector the same length as the number
of points in the plot (9 in our case, as there are 9 samples). Each element of
the vector should be a colour name (R understands
<a href="http://www.stat.columbia.edu/~tzheng/files/Rcolor.pdf">over 600 colour names</a>),
indicating what colour that sample should be.</p>
<p>To make this more clear, take a look at the table below, which lists all the
samples and the phenotype group they belong to:</p>
<table class="table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
Samples
</th>
<th style="text-align:left;">
Group
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
10_6_5_11
</td>
<td style="text-align:left;">
LP
</td>
</tr>
<tr>
<td style="text-align:left;">
9_6_5_11
</td>
<td style="text-align:left;">
ML
</td>
</tr>
<tr>
<td style="text-align:left;">
purep53
</td>
<td style="text-align:left;">
Basal
</td>
</tr>
<tr>
<td style="text-align:left;">
JMS8-2
</td>
<td style="text-align:left;">
Basal
</td>
</tr>
<tr>
<td style="text-align:left;">
JMS8-3
</td>
<td style="text-align:left;">
ML
</td>
</tr>
<tr>
<td style="text-align:left;">
JMS8-4
</td>
<td style="text-align:left;">
LP
</td>
</tr>
<tr>
<td style="text-align:left;">
JMS8-5
</td>
<td style="text-align:left;">
Basal
</td>
</tr>
<tr>
<td style="text-align:left;">
JMS9-P7c
</td>
<td style="text-align:left;">
ML
</td>
</tr>
<tr>
<td style="text-align:left;">
JMS9-P8c
</td>
<td style="text-align:left;">
LP
</td>
</tr>
</tbody>
</table>
<p>For example, let’s say we wanted LP samples to be coloured green, ML samples to
be coloured red and Basal samples to be coloured blue. The <code>col</code> argument would
then require a vector that we can generate as follows</p>
<div class="sourceCode" id="cb250"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb250-1"><a href="session-4.html#cb250-1"></a>group_col &lt;-<span class="st"> </span>dge<span class="op">$</span>samples<span class="op">$</span>group</span>
<span id="cb250-2"><a href="session-4.html#cb250-2"></a><span class="kw">levels</span>(group_col) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;green&quot;</span>, <span class="st">&quot;red&quot;</span>)</span>
<span id="cb250-3"><a href="session-4.html#cb250-3"></a>group_col &lt;-<span class="st"> </span><span class="kw">as.character</span>(group_col)</span>
<span id="cb250-4"><a href="session-4.html#cb250-4"></a>group_col</span></code></pre></div>
<pre><code>## [1] &quot;green&quot; &quot;red&quot;   &quot;blue&quot;  &quot;blue&quot;  &quot;red&quot;   &quot;green&quot; &quot;blue&quot;  &quot;red&quot;   &quot;green&quot;</code></pre>
<p>We can also add a legend to the figure by running the <code>legend()</code> function
immediately after a new figure is created. We have to specify where to position
the legend as well as the labels and colours within the legend.</p>
<div class="sourceCode" id="cb252"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb252-1"><a href="session-4.html#cb252-1"></a><span class="kw">plotMDS</span>(dge, <span class="dt">labels =</span> group, <span class="dt">col =</span> group_col)</span>
<span id="cb252-2"><a href="session-4.html#cb252-2"></a><span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>, <span class="dt">legend =</span> <span class="kw">c</span>(<span class="st">&quot;Basal&quot;</span>, <span class="st">&quot;LP&quot;</span>, <span class="st">&quot;ML&quot;</span>), <span class="dt">fill =</span> <span class="kw">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;green&quot;</span>, <span class="st">&quot;red&quot;</span>))</span></code></pre></div>
<p><img src="IntroBaseR_files/figure-html/unnamed-chunk-118-1.png" width="672" /></p>
</div>
<div id="linear-modelling" class="section level2">
<h2><span class="header-section-number">4.5</span> Linear modelling</h2>
<p>The next step of the limma-voom analysis is to fit a linear model for each gene.
A linear model is a broad class of statistical models that predict a variable of
interest using one or more ‘explanatory’ (also called ‘predictor’) variables.
The most basic type is linear regression, which models the relationship between
a continuous variable and continuous or categorical ‘explanatory’ variables. It
uses an equation that looks like this:</p>
<p><span class="math inline">\(Y = \beta_{0} + \beta_{1}X_{1} + \beta_{2}X_{2}...\)</span></p>
<p>This equation is saying that a response variable of interest <span class="math inline">\(Y\)</span> is equal to a
constant (<span class="math inline">\(\beta_{0}\)</span>) plus the sum of the covariates (<span class="math inline">\(X_{i}\)</span>) each multiplied
by a constant coefficient (<span class="math inline">\(\beta_{i}\)</span>).</p>
<p>Our experiment is quite simple, since there is only a single covariate, the
cell type. The true benefit of using linear models is in its ability to
accommodate more complex designs including multiple covariates.</p>
<p>To fit the linear models in the limma-voom framework we need two objects in
addition to our data:</p>
<ul>
<li>A design matrix, representing the covariates.</li>
<li>A contrast matrix, representing the specific comparison we wish to make.</li>
</ul>
<div id="design-matrix" class="section level3">
<h3><span class="header-section-number">4.5.1</span> Design matrix</h3>
<p>The design matrix specifies the values of the covariates for each sample. This
is represented as a matrix due to the mathematical convenience.</p>
<p>To generate a design matrix. We use the function <code>model.matrix()</code>, with the
expression <code>~0 + group</code>. This returns a matrix representing the design where
there is no intercept term and group is the only covariate. If we omit the
<code>0</code> then there would be an intercept in the model, and if we included more
covariates then more columns would be generated.</p>
<div class="sourceCode" id="cb253"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb253-1"><a href="session-4.html#cb253-1"></a>design &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span><span class="dv">0</span> <span class="op">+</span><span class="st"> </span>group, <span class="dt">data =</span> dge<span class="op">$</span>samples)</span>
<span id="cb253-2"><a href="session-4.html#cb253-2"></a>design</span></code></pre></div>
<pre><code>##           groupBasal groupLP groupML
## 10_6_5_11          0       1       0
## 9_6_5_11           0       0       1
## purep53            1       0       0
## JMS8-2             1       0       0
## JMS8-3             0       0       1
## JMS8-4             0       1       0
## JMS8-5             1       0       0
## JMS9-P7c           0       0       1
## JMS9-P8c           0       1       0
## attr(,&quot;assign&quot;)
## [1] 1 1 1
## attr(,&quot;contrasts&quot;)
## attr(,&quot;contrasts&quot;)$group
## [1] &quot;contr.treatment&quot;</code></pre>
<p>There are 9 rows, one for each sample. Along the columns are the names of the
groups. The values in the cells denote membership of the particular sample
for a particular group, as our groups in this case are mutually exclusive,
each row contains only a single 1 to denote membership in a single group.</p>
</div>
<div id="contrasts" class="section level3">
<h3><span class="header-section-number">4.5.2</span> Contrasts</h3>
<p>‘Contrasts’ let us ask specific questions between our experimental groups. In
our data we have 3 experimental groups, if we are to test for differential
expression, we are most likely interested in differences between only two of the
groups at a time. Contrasts let us specify exactly what we’re testing for, and
is also represented by a matrix just like the design.</p>
<p>A contrast matrix can be made using the <code>makeContrasts()</code> function. Within this
function, we specify the name of each specific contrast and the formula for that
contrast. For example, the <code>BasalvsLP</code> contrasts compares the difference between
the <code>Basal</code> and <code>LP</code> groups. Note that the name of the phenotype groups must be
written exactly as they are in the column names of our design matrix (see
above).</p>
<p>In addition to the individual contrasts, the function must know about the design
of the model. This is passed through the <code>levels</code> argument, which either accepts
a matrix with the column names corresponding to levels of your experimental
groups, or the levels themselves as a character vector</p>
<div class="sourceCode" id="cb255"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb255-1"><a href="session-4.html#cb255-1"></a>contr.matrix &lt;-<span class="st"> </span><span class="kw">makeContrasts</span>(<span class="dt">BasalvsLP =</span> <span class="st">&quot;Basal - LP&quot;</span>, <span class="dt">BasalvsML =</span> <span class="st">&quot;Basal - ML&quot;</span>, </span>
<span id="cb255-2"><a href="session-4.html#cb255-2"></a>    <span class="dt">LPvsML =</span> <span class="st">&quot;LP - ML&quot;</span>, <span class="dt">levels =</span> design)  <span class="co"># alternatively &#39;levels = colnames(design)&#39;</span></span>
<span id="cb255-3"><a href="session-4.html#cb255-3"></a>contr.matrix</span></code></pre></div>
<pre><code>##        Contrasts
## Levels  BasalvsLP BasalvsML LPvsML
##   Basal         1         1      0
##   LP           -1         0      1
##   ML            0        -1     -1</code></pre>
<p>Note that the sum of all the numbers along each column is 0. The first column
is the contrast for the difference between Basal (1) and LP (-1). This property
is required for valid contrast matrices. An alternative test may be between
one group and the average of the others which would look like <code>c(1, -0.5, -0.5)</code>
down one of the columns.</p>
</div>
<div id="variance-modelling-with-voom" class="section level3">
<h3><span class="header-section-number">4.5.3</span> Variance modelling with voom</h3>
<p>We are now ready to fit our linear models. Limma fits linear models to the data
with the assumption that the underlying data is normally distributed. Count data
is generally not normally distributed, but log transforming count data gives it
a roughly normal distribution sufficient for linear models to work well. To do
this, limma transforms the raw count data to log-cpm using library sizes
and the normalisation factors we calculated previously.</p>
<p>In addition to the normalisation steps, the limma-voom pipeline uses the
<code>voom()</code> function to generate weights for the individual genes based on a
modelled mean-variance relationship. This modelling allows use to get more
information out of small sample sizes as the weights prevent our model from
being more heavily influenced by more variable data points.</p>
<p>The <code>voom()</code> function takes as arguments, our <code>DGEList</code> object and our design
matrix. It also optionally outputs a plot of the mean-variance relationship of
our data, called the ‘voom-plot’.</p>
<div class="sourceCode" id="cb257"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb257-1"><a href="session-4.html#cb257-1"></a>v &lt;-<span class="st"> </span><span class="kw">voom</span>(dge, design, <span class="dt">plot =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<p><img src="IntroBaseR_files/figure-html/unnamed-chunk-122-1.png" width="672" /></p>
<p>The output of <code>voom()</code> (our variable <code>v</code>) is an <code>EList</code> object which contains
the following elements:</p>
<ul>
<li><code>genes</code> - a data frame of gene annotation data.</li>
<li><code>targets</code> - data frame of sample data.</li>
<li><code>E</code> - numeric matrix of normalised log-cpm values.</li>
<li><code>weights</code> - numeric matrix of precision weights.</li>
<li><code>design</code> - the design matrix.</li>
</ul>
<p><strong>Challenge 4.2</strong></p>
<ol style="list-style-type: decimal">
<li><p>Create the MDS plot with legend, but with solid dots instead of text within
the plot.</p></li>
<li><p>Which of the elements of the <code>EList</code> is newly generated, which are taken
from previous objects?</p></li>
</ol>
</div>
<div id="fitting-the-linear-model" class="section level3">
<h3><span class="header-section-number">4.5.4</span> Fitting the linear model</h3>
<p>We are now ready to fit our linear model with <code>lmFit()</code>, which calculates
coefficients we defined in our design matrix <code>design</code>. The resulting object,
<code>vfit</code> is a <code>MArrayLM</code> object. It contains a information about our genes (the
same data frame as <code>genes</code> from our <code>EList</code> object <code>v</code> above), the design matrix
and a number of statistical outputs. Of most interest to us is the coefficients,
stored in an element called <code>coefficients</code>. The first rows of this matrix is
shown below. Each gene is row and is labelled using the EntrezID. Each column
gives coefficients for each of our phenotype groups. These coefficients are
weighted averages of the log-cpm of each gene in each group.</p>
<div class="sourceCode" id="cb258"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb258-1"><a href="session-4.html#cb258-1"></a>vfit &lt;-<span class="st"> </span><span class="kw">lmFit</span>(v, design)</span>
<span id="cb258-2"><a href="session-4.html#cb258-2"></a><span class="kw">head</span>(vfit<span class="op">$</span>coefficients)</span></code></pre></div>
<pre><code>##             Basal        LP        ML
## 497097  3.0241632 -4.490392 -3.944477
## 20671   0.2681245 -2.488746 -2.024896
## 27395   4.3271126  3.901078  4.365378
## 18777   5.2069566  4.976083  5.654066
## 21399   5.2108711  4.901842  4.876380
## 58175  -1.9296994  3.581328  3.133985</code></pre>
<p>We can then use <code>contrasts.fit()</code> to calculate coefficients for each contrast
(or ‘comparison’) we specified in our <code>contr.matrix</code>. The output is also an
object of the class <code>MArrayLM</code> (also known as an <code>MArrayLM</code> object). When we
inspect the <code>coefficients</code> element now, we can see that each column is a
contrast that we specified in our contrast matrix.</p>
<div class="sourceCode" id="cb260"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb260-1"><a href="session-4.html#cb260-1"></a>vfit &lt;-<span class="st"> </span><span class="kw">contrasts.fit</span>(vfit, <span class="dt">contrasts =</span> contr.matrix)</span>
<span id="cb260-2"><a href="session-4.html#cb260-2"></a><span class="kw">head</span>(vfit<span class="op">$</span>coefficients)</span></code></pre></div>
<pre><code>##         Contrasts
##           BasalvsLP   BasalvsML      LPvsML
##   497097  7.5145557  6.96864007 -0.54591559
##   20671   2.7568708  2.29302094 -0.46384989
##   27395   0.4260347 -0.03826548 -0.46430022
##   18777   0.2308732 -0.44710891 -0.67798213
##   21399   0.3090294  0.33449065  0.02546125
##   58175  -5.5110274 -5.06368468  0.44734272</code></pre>
</div>
</div>
<div id="statistical-testing" class="section level2">
<h2><span class="header-section-number">4.6</span> Statistical testing</h2>
<p>The next step is to carry out statistical testing to determine which genes are
differentially expressed. The function <code>eBayes()</code> computes moderated
t-statistics, moderated F-statistics and log-odds of differential expression for
each gene, given a fitted linear model. ‘Moderated’ refers to empirical Bayes
moderation, which borrows information across genes to obtain more accurate
measures of variability for each gene. This also increases our power to detect
differentially expressed genes.</p>
<div class="sourceCode" id="cb262"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb262-1"><a href="session-4.html#cb262-1"></a>efit &lt;-<span class="st"> </span><span class="kw">eBayes</span>(vfit)</span></code></pre></div>
<p>We can now look at the number of differentially expressed genes using the
<code>decideTests()</code> function. The output of this function is a matrix where each
column is a contrast (comparison of interest) and each row is a gene. The
numbers 1, -1 and 0 mean up-regulated, down-regulated or not significantly
differentially expressed, respectively.</p>
<p>Note that <code>decideTests()</code> also accounts for multiple testing. The default method
is Benjamini and Hochberg <span class="citation">(Benjamini and Hochberg <a href="#ref-benjamini1995controlling" role="doc-biblioref">1995</a>)</span> but several others are also
available.</p>
<div class="sourceCode" id="cb263"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb263-1"><a href="session-4.html#cb263-1"></a>dt &lt;-<span class="st"> </span><span class="kw">decideTests</span>(efit)</span>
<span id="cb263-2"><a href="session-4.html#cb263-2"></a>dt</span></code></pre></div>
<pre><code>## TestResults matrix
##         Contrasts
##          BasalvsLP BasalvsML LPvsML
##   497097         1         1      0
##   20671          1         1      0
##   27395          0         0      0
##   18777          0        -1     -1
##   21399          0         1      0
## 16619 more rows ...</code></pre>
<p>To obtain the total number of differentially expressed genes for each
comparison, we can add the function <code>summary()</code>:</p>
<div class="sourceCode" id="cb265"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb265-1"><a href="session-4.html#cb265-1"></a><span class="kw">summary</span>(dt)</span></code></pre></div>
<pre><code>##        BasalvsLP BasalvsML LPvsML
## Down        4500      4850   2701
## NotSig      7307      6996  11821
## Up          4817      4778   2102</code></pre>
<p>The function <code>topTable()</code> can be used to obtain more information on the
differentially expressed genes for each contrast. <code>topTable()</code> takes as
arguments the <code>MArrayLM</code> object output by <code>eBayes()</code> (<code>efit</code>), the contrast name
of interest and the number of top differentially expressed genes to output. Note
that the contrast name must be given in quotes and must be exactly as written in
the contrast matrix <code>contr.matrix</code>.</p>
<p>It outputs a data frame with the following information:</p>
<ul>
<li><strong>Gene details</strong> - gene information, from the <code>gene</code> element of the <code>MArrayLM</code>
object (<code>efit</code>).</li>
<li><code>logFC</code> - the log2 fold change of the contrast.</li>
<li><code>AveExpr</code> - the average log2 expression of that gene.</li>
<li><code>t</code> - moderated t-statistic.</li>
<li><code>P.Value</code> - p value.</li>
<li><code>adj.P.Val</code> - adjusted p value.</li>
<li><code>B</code> - log-odds that the gene is differentially expressed.</li>
</ul>
<div class="sourceCode" id="cb267"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb267-1"><a href="session-4.html#cb267-1"></a>top &lt;-<span class="st"> </span><span class="kw">topTable</span>(efit, <span class="dt">coef =</span> <span class="st">&quot;BasalvsLP&quot;</span>, <span class="dt">n =</span> <span class="ot">Inf</span>)</span>
<span id="cb267-2"><a href="session-4.html#cb267-2"></a><span class="kw">head</span>(top)</span></code></pre></div>
<pre><code>##        ENTREZID   SYMBOL TXCHROM     logFC  AveExpr         t      P.Value
## 12521     12521     Cd82    chr2 -4.095479 7.069637 -35.46821 3.383539e-12
## 22249     22249   Unc13b    chr4 -4.350553 5.663171 -32.38571 8.648521e-12
## 16324     16324    Inhbb    chr1 -4.721417 6.460922 -30.80826 1.447137e-11
## 14245     14245    Lpin1   chr12 -3.768977 6.294017 -29.94805 1.937313e-11
## 218518   218518 Marveld2   chr13 -5.215232 4.930008 -30.87689 1.414332e-11
## 12759     12759      Clu   chr14 -5.306847 8.856581 -29.55253 2.221529e-11
##           adj.P.Val        B
## 12521  4.660973e-08 18.41326
## 22249  4.660973e-08 17.42652
## 16324  4.660973e-08 17.06291
## 14245  4.660973e-08 16.85869
## 218518 4.660973e-08 16.76959
## 12759  4.660973e-08 16.70221</code></pre>
<p>With that we can complete our analysis by writing out some results</p>
<div class="sourceCode" id="cb269"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb269-1"><a href="session-4.html#cb269-1"></a><span class="kw">write.csv</span>(top, <span class="dt">file =</span> <span class="st">&quot;BasalvsLP.csv&quot;</span>)</span></code></pre></div>
</div>
<div id="ma-plot" class="section level2">
<h2><span class="header-section-number">4.7</span> MA Plot</h2>
<p>The MA plot is a plot of log-fold-change (M-values) against log-expression
averages (A-values), this is a common plot in RNA sequencing analysis to
visualise the result of differential expression tests. It can be created
using the <code>plotMA()</code> from the <code>limma</code> package. Creating this plot requires 3
pieces of information:</p>
<ul>
<li><code>object = efit</code>: The the fitted object containing the log-fold-change and log-expression
averages</li>
<li><code>coef = 1</code>: The column number of the contrast to plot since there are 3
different contrasts fitted within the object.</li>
<li><code>status = dt[, 1]</code>: A vector of numerics denoting whether a gene is
up-regulated or down-regulated.</li>
</ul>
<div class="sourceCode" id="cb270"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb270-1"><a href="session-4.html#cb270-1"></a><span class="kw">plotMA</span>(efit, <span class="dt">coef =</span> <span class="dv">1</span>, <span class="dt">status =</span> dt[, <span class="st">&quot;BasalvsLP&quot;</span>])</span></code></pre></div>
<p><img src="IntroBaseR_files/figure-html/unnamed-chunk-130-1.png" width="672" /></p>
<p>We can also save this plot programmatically as a PDF for further editing. To do
this we use <code>pdf()</code> to turn on the pdf capture device, run the command that
creates the plot, which is now captured by the pdf, and then turn the device
off.</p>
<div class="sourceCode" id="cb271"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb271-1"><a href="session-4.html#cb271-1"></a><span class="kw">pdf</span>(<span class="dt">file =</span> <span class="st">&quot;BasalvsLP-MAPlot.pdf&quot;</span>)</span>
<span id="cb271-2"><a href="session-4.html#cb271-2"></a><span class="kw">plotMA</span>(efit, <span class="dt">coef =</span> <span class="dv">1</span>, <span class="dt">status =</span> dt[, <span class="st">&quot;BasalvsLP&quot;</span>])</span>
<span id="cb271-3"><a href="session-4.html#cb271-3"></a><span class="kw">dev.off</span>()</span></code></pre></div>
<p><strong>Challenge 4.3</strong></p>
<ol style="list-style-type: decimal">
<li><p>What are the symbols of the top 5 most differentially expressed genes
between basal and luminal progenitor samples in chromosome X based on adjusted
p-value?</p></li>
<li><p>Create a volcano (scatter) plot of log-fold-change on the x-axis and
-log10(p-value) on y-axis.</p></li>
</ol>

</div>
</div>
<h3>References</h3>
<div id="refs" class="references">
<div id="ref-benjamini1995controlling">
<p>Benjamini, Yoav, and Yosef Hochberg. 1995. “Controlling the False Discovery Rate: A Practical and Powerful Approach to Multiple Testing.” <em>Journal of the Royal Statistical Society: Series B (Methodological)</em> 57 (1): 289–300.</p>
</div>
<div id="ref-bourgon2010independent">
<p>Bourgon, Richard, Robert Gentleman, and Wolfgang Huber. 2010. “Independent Filtering Increases Detection Power for High-Throughput Experiments.” <em>Proceedings of the National Academy of Sciences</em> 107 (21): 9546–51.</p>
</div>
<div id="ref-mcintyre2011rna">
<p>McIntyre, Lauren M, Kenneth K Lopiano, Alison M Morse, Victor Amin, Ann L Oberg, Linda J Young, and Sergey V Nuzhdin. 2011. “RNA-Seq: Technical Variability and Sampling.” <em>BMC Genomics</em> 12 (1): 293.</p>
</div>
<div id="ref-robinson2010scaling">
<p>Robinson, Mark D, and Alicia Oshlack. 2010. “A Scaling Normalization Method for Differential Expression Analysis of Rna-Seq Data.” <em>Genome Biology</em> 11 (3): R25.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="session-3.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="further-reading.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/04-session4.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["IntroBaseR.pdf", "IntroBaseR.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
